File: app/src/main/java/com/mirash/familiar/activity/pin/PinCodeConfirmActivity.java
package com.mirash.familiar.activity.pin;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.CallSuper;

import com.mirash.familiar.Const;
import com.mirash.familiar.R;
import com.mirash.familiar.activity.main.MainActivity;
import com.mirash.familiar.preferences.EncryptedAppPreferences;
import com.mirash.familiar.tool.Utils;

import java.util.Objects;

/**
 * @author Mirash
 */
public class PinCodeConfirmActivity extends PinCodeBaseActivity {

    private String desiredPinCode;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        desiredPinCode = getIntent().getStringExtra(Const.KEY_PIN_CODE);
        if (desiredPinCode == null || desiredPinCode.length() != Const.PIN_CODE_SIZE) {
            finishAffinity();
        }
        buttonBottomStart.setVisibility(View.GONE);
        buttonBottomEnd.setVisibility(View.GONE);
        pinCodeDetailsView.setAlpha(0);
        pinCodeDetailsView.setVisibility(View.VISIBLE);
        pinCodeDetailsView.setText(getString(R.string.pin_code_incorrect));
    }

    @Override
    protected int getMessageRes() {
        return R.string.pin_code_confirm;
    }

    @Override
    protected void checkPinCode(String pinCode) {
        if (pinCode == null || pinCode.length() != Const.PIN_CODE_SIZE) return;
        if (Objects.equals(desiredPinCode, pinCode)) {
            EncryptedAppPreferences.getInstance().setPinCode(pinCode);
            startNewActivity(MainActivity.class);
        } else {
            changeAlpha(pinCodeDetailsView, 1);
        }
    }

    @CallSuper
    protected void handleEnteredPinCodeSizeChange(int size) {
        super.handleEnteredPinCodeSizeChange(size);
        if (size < Const.PIN_CODE_SIZE) {
            changeAlpha(pinCodeDetailsView, 0);
        }
    }

    private void changeAlpha(View view, float to) {
        float from = view.getAlpha();
        if (from == to) return;
        int duration = (int) Math.max(1, Math.abs(to - from) * 250);
        view.animate().setDuration(duration).alpha(to).start();
    }

    @Override
    public void onBackPressed() {
        if (Utils.isPinCodeActual(EncryptedAppPreferences.getInstance().getPinCode())) {
            startNewActivity(PinCodeEnterActivity.class);
        } else {
            onBackPressed();
        }
    }
}

File: app/src/main/java/com/mirash/familiar/activity/pin/PinCodeConfirmActivity.kt
package com.mirash.familiar.activity.pin

import android.os.Bundle
import android.view.View
import androidx.activity.OnBackPressedCallback
import androidx.annotation.CallSuper
import com.mirash.familiar.R
import com.mirash.familiar.activity.main.MainActivity
import com.mirash.familiar.preferences.EncryptedAppPreferences
import com.mirash.familiar.tool.KEY_PIN_CODE
import com.mirash.familiar.tool.PIN_CODE_SIZE
import kotlin.math.abs
import kotlin.math.max

/**
 * @author Mirash
 */
class PinCodeConfirmActivity : PinCodeBaseActivity() {
    private var desiredPinCode: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        intent.getStringExtra(KEY_PIN_CODE).let {
            if (it == null || it.length != PIN_CODE_SIZE) {
                finishAffinity()
            }
            desiredPinCode = it
        }
        binding.pinButtonBottomStart.visibility = View.GONE
        binding.pinButtonBottomEnd.visibility = View.GONE
        binding.pinCodeDetails.let {
            it.setAlpha(0f)
            it.visibility = View.VISIBLE
            it.text = getString(R.string.pin_code_incorrect)
        }
        onBackPressedDispatcher.addCallback(object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                startNewFinishCurrentActivity(PinCodeEnterActivity::class.java)
            }
        })
    }

    override fun getMessageRes(): Int = R.string.pin_code_confirm

    override fun checkPinCode(pinCode: String?) {
        if (pinCode == null || pinCode.length != PIN_CODE_SIZE) return
        if (desiredPinCode == pinCode) {
            EncryptedAppPreferences.pinCode = pinCode
            startNewFinishCurrentActivity(MainActivity::class.java)
        } else {
            changeAlpha(binding.pinCodeDetails, 1f)
        }
    }

    @CallSuper
    override fun handleEnteredPinCodeSizeChange(size: Int) {
        super.handleEnteredPinCodeSizeChange(size)
        if (size < PIN_CODE_SIZE) {
            changeAlpha(binding.pinCodeDetails, 0f)
        }
    }

    private fun changeAlpha(view: View, to: Float) {
        val from = view.alpha
        if (from == to) return
        val duration = max(1.0, (abs((to - from).toDouble()) * 250)).toInt()
        view.animate().setDuration(duration.toLong()).alpha(to).start()
    }
}